riscv_ctg/env/arch_test.h

```diff
@@ -575,37 +576,81 @@
 #define NARG(...) _ARG5(__VA_OPT__(__VA_ARGS__,)4,3,2,1,0)
 #define RVTEST_SIGUPD(_BR,_R,...)\
   .if NARG(__VA_ARGS__) == 1;\
-    SREG _R,_ARG1(__VA_ARGS__,0)(_BR);\
+    .if XLEN==64 && FLEN==32;\
+      sw _R,_ARG1(__VA_ARGS__,0)(_BR);\
+    .else;\
+      SREG _R,_ARG1(__VA_ARGS__,0)(_BR);\
+    .endif;\
     .set offset,_ARG1(__VA_OPT__(__VA_ARGS__,)0)+REGWIDTH;\
   .endif;\
   .if NARG(__VA_ARGS__) == 0;\
-    SREG _R,offset(_BR);\
-  .set offset,offset+REGWIDTH;\
+    .if XLEN==64 && FLEN==32;\
+      sw _R,offset(_BR);\
+    .else;\
+      SREG _R,offset(_BR);\
+    .endif;\
+    .set offset,offset+REGWIDTH;\
   .endif;
   
-#define RVTEST_SIGUPD_F(_BR,_R,_F,...)\
-  .if NARG(__VA_ARGS__) == 1;\
-    FSREG _R,_ARG1(__VA_ARGS__,0)(_BR);\
-    SREG _F,_ARG1(__VA_ARGS__,0)+REGWIDTH(_BR);\
-    .set offset,_ARG1(__VA_OPT__(__VA_ARGS__,)0)+(REGWIDTH+REGWIDTH);\
-  .endif;\
-  .if NARG(__VA_ARGS__) == 0;\
-    FSREG _R,offset(_BR);\
-    SREG _F,offset+REGWIDTH(_BR);\
-    .set offset,offset+(REGWIDTH+REGWIDTH);\
-  .endif;
+  
+#define RVTEST_SIGUPD_F(_BR,_R,_F,...) \
+  FLENN() ;\
+  .if NARG(__VA_ARGS__) == 1	;\
+    .set offset, _ARG1(__VA_ARGS__,0) ;\
+    .if offset > 2048-(2*SIGALIGN) ;\
+      addi _BR, _BR, 2048-(2*SIGALIGN) ;\
+      .set offset, offset-(2048-(2*SIGALIGN)) ;\
+    .endif  ;\
+    .if (offset&(SIGALIGN-1))!=0 ;\
+       .warning "Incorrect Offset Alginment for signature." ;\
+    .endif;\
+    FSREG _R,offset+ 0(_BR) ;\
+    .if XLEN==64 && FLEN==32;\
+      sw _F,offset+SIGALIGN(_BR) ;\
+    .else;\
+      SREG _F,offset+SIGALIGN(_BR) ;\
+    .endif;\
+    .set offset,offset+(2*SIGALIGN) ;\
+  .endif ;
+
   
 #define RVTEST_SIGUPD_FID(_BR,_R,_F,...)\
+  FLENN();\
   .if NARG(__VA_ARGS__) == 1;\
-    SREG _R,_ARG1(__VA_ARGS__,0)(_BR);\
-    SREG _F,_ARG1(__VA_ARGS__,0)+REGWIDTH(_BR);\
+    .if XLEN==64 && FLEN==32;\
+      sw _R,_ARG1(__VA_ARGS__,0)(_BR);\
+      sw _F,_ARG1(__VA_ARGS__,0)+REGWIDTH(_BR);\
```

Comment:
Should not be `+REGWIDTH`. Should be `+8`.

Reply:
For XLEN==64 && FLEN==32 the regwidth should be 4 right?